File: baseclient/ts/src/baseclient.ts
import * as $tea from '@alicloud/tea-typescript';
import Creadential from '@alicloud/credentials';
import { Parser, Builder } from 'xml2js';
import kitx from 'kitx';
import { platform, arch } from 'os';
import { getType } from 'mime';
import { Readable } from 'stream';
import { createHash } from 'crypto';
import Crc64 from './crc';
import { getSignatureV2, getSignatureV1 } from './signature';

function parseXML(body: string): any {
  let parser = new Parser({ explicitArray: false });
  let result: {[key: string]: any} = {};
  // parseString 实际上是一个同步方法
  parser.parseString(body, function (err: any, output: any) {
    result.err = err;
    result.output = output;
  });

  if (result.err) {
    throw result.err;
  }

  return result.output;
}

export default class BaseClient {

  _endpoint: string
  _regionId: string
  _autoretry: boolean
  _protocol: string
  _userAgent: string
  _readTimeout: number
  _connectTimeout: number
  _localAddr: string
  _httpProxy: string
  _httpsProxy: string
  _hostModel: string
  _noProxy: string
  _socks5Proxy: string
  _socks5NetWork: string
  _isEnableMD5: boolean
  _isEnableCrc: boolean
  _signatureVersion: string
  _addtionalHeaders: string[]
  _maxIdleConns: number
  _logger: any
  _creadentials: Creadential
  _contentMd5: string
  _contentLength: number

  constructor(config: { [key: string]: any }) {
    this._protocol = config['protocol'];
    this._endpoint = config['endpoint'];
    this._regionId = config['regionId'] || 'cn-hangzhou';
    this._signatureVersion = 'v1';
    this._isEnableMD5 = config['isEnableMD5'] || false;
    this._isEnableCrc = config['isEnableCrc'] || false;
    this._userAgent = `AlibabaCloud (${platform()}; ${arch()}) Node.js/${process.version} Core/1.0.1`
    this._hostModel = config['hostModel'] || 'domain';

    this._addtionalHeaders = [];
    if (!config['type']) {
      config['type'] = 'access_key';
    }

    this._creadentials = new Creadential({
      accessKeyId: config['accessKeyId'],
      accessKeySecret: config['accessKeySecret'],
      securityToken: config['securityToken'],
      type: config['type'],
    }, {});
  }

  _addAddtionalHeaders(header: string): void {
    this._addtionalHeaders.push(header);
  }

  _setAddtionalHeaders(headers: string[]): void {
    if (!headers) {
      return;
    }
    this._addtionalHeaders = headers;
  }

  _getAddtionalHeaders(): string[] {
    return this._addtionalHeaders;
  }
  _setSignatureVersion(version: string): void {
    if (!version) {
      return;
    }
    this._signatureVersion = version.toLocaleLowerCase();
  }

  _setHostModel(model: string) {
    if (!model) {
      return;
    }
    this._hostModel = model.toLocaleLowerCase();
  }

  _setRegionId(regionId: string) {
    this._regionId = regionId || 'cn-hangzhou';
  }

  _getDate(): string {
    let now = new Date();
    return now.toUTCString();
  }

  _xmlCast<T>(obj: any, clazz: T): { [key: string]: any } {
    obj = obj || {};
    let ret: { [key: string]: any } = {};
    let clz = clazz as any;
    let names: { [key: string]: string } = clz.names();
    let types: { [key: string]: any } = clz.types();

    Object.keys(names).forEach((key) => {
      let originName = names[key];
      let value = obj[originName];
      let type = types[key];
      switch (type) {
        case 'boolean':
          if (!value) {
            ret[originName] = false;
            return;
          }
          ret[originName] = value === 'false' ? false : true;
          return;
        case 'number':
          if (value != 0 && !value) {
            ret[originName] = NaN;
            return;
          }
          ret[originName] = +value;
          return;
        case 'string':
          if (!value) {
            ret[originName] = '';
            return;
          }
          ret[originName] = value.toString();
          return;
        default:
          if (type.type === 'array') {
            if (!value) {
              ret[originName] = [];
              return;
            }
            if (!Array.isArray(value)) {
              value = [value];
            }
            if (typeof type.itemType === 'function') {
              ret[originName] = value.map((d: any) => {
                return this._xmlCast(d, type.itemType);
              });
            } else {
              ret[originName] = value;
            }
          } else if (typeof type === 'function') {
            if (!value) {
              value = {}
            }
            ret[originName] = this._xmlCast(value, type);
          } else {
            ret[originName] = value;
          }
      }
    })
    return ret;
  }

  async _readAsString(response: $tea.Response): Promise<string> {
    let bytes = await response.readBytes();
    return bytes.toString();
  }

  _parseXml<T>(body: string, clazz: T): { [key: string]: any } {
    let ret: { [key: string]: any } = parseXML(body);
    if (typeof clazz !== 'undefined') {
      ret = this._xmlCast(ret, clazz);
    }
    return ret;
  }

  _equal(a: string, b: string): boolean {
    return a === b;
  }

  _empty(input: string): boolean {
    return !input;
  }

  _ifListEmpty(list: string[]): boolean {
    return !list || list.length === 0
  }

  _listToString(list: string[], separator: string) : string {
    return list.join(separator);
  }

  _getSignatureV1(request: $tea.Request, bucketName: string, accessKeySecret: string): string {
    let result = getSignatureV1(request, bucketName, accessKeySecret);
    return result.signature;
  }

  _getSignatureV2(request: $tea.Request, bucketName: string, accessKeySecret: string, addtionalHeaders: string[]): string {
    let result = getSignatureV2(request, bucketName, accessKeySecret, addtionalHeaders);
    return result.signature;
  }

  _default(value: string, defaultVal: string): string {
    if (!value) {
      return defaultVal;
    }
    return value;
  }

  _defaultNumber(number: number, defaultNum: number): number {
    if (!number) {
      return defaultNum;
    }
    return number;
  }

  _isFail(response: { [key: string]: any }): boolean {
    if (!response) {
      return true;
    }
    if (response.statusCode >= 200 && response.statusCode < 300) {
      return false;
    }
    return true;
  }

  _toQuery(query: { [key: string]: any }): { [key: string]: string } {
    let ret: { [key: string]: string } = {};
    if (!query) {
      return ret;
    }
    for (let [key, value] of Object.entries(query)) {
      if (typeof value === 'undefined' || value == null) {
        ret[key] = '';
        continue;
      }
      ret[key] = value.toString();
    }
    return ret;
  }

  _toMeta(meta: { [key: string]: any }, prefix: string): { [key: string]: string } {
    let ret: { [key: string]: string } = {};
    if (!meta) {
      return ret;
    }
    Object.keys(meta).forEach((key) => {
      if (typeof meta[key] === 'undefined' || meta[key] == null){
        meta[key] = '';
      }
      if (!key.startsWith(prefix)) {
        ret[prefix + key] = meta[key].toString();
        return;
      }
      ret[key] = meta[key].toString();
    });
    return ret;
  }

  _parseMeta(meta: { [key: string]: any }, prefix: string): { [key: string]: string } {
    let ret: { [key: string]: string } = {};
    if (!meta) {
      return ret;
    }
    Object.keys(meta).forEach((key) => {
      if (key.startsWith(prefix)) {
        ret[key.replace(prefix, '')] = meta[key];
        return;
      }
      ret[key] = meta[key];
    });
    return ret;
  }

  _getContentMD5(body: string): string {
    return kitx.md5(body, 'hex');
  }

  _getContentType(name: string): string {
    return getType(name) || 'text/plain';
  }

  _isNotCrcMatched(calCrc: Buffer, respCrc: string): boolean {
    if (this._isEnableCrc && calCrc.toString() !== respCrc) {
      return true;
    }
    return false;
  }

  _encode(value: string, type: string): string {
    if (!value) {
      return '';
    }
    let strs = value.split('/');
    let str = strs[strs.length - 1];
    switch (type) {
      case 'UrlEncode':
        str = encodeURIComponent(str);
        strs[strs.length - 1] = str;
        return strs.join('/');
      case 'Base64':
        str = Buffer.from(str).toString('base64');
        strs[strs.length - 1] = str;
        return strs.join('/');
    }
  }

  _base64Decode(value: string): string {
    if (!value) {
      return '';
    }
    const strs = value.split('/');
    const result = Buffer.from(strs[strs.length - 1], 'base64').toString('utf8');
    strs[strs.length - 1] = result
    return strs.join('/');
  }

  _getCrc(a: $tea.Request, b: string, c: any, d: any): Buffer {
    throw new Error('the method is un-implemented!');
  }

  _getRespCrc(response: $tea.Response, b: boolean, c: any, d: any): Buffer {
    throw new Error('the method is un-implemented!');
  }

  _getUserAgent(userAgent: string = ''): string {
    if (userAgent) {
      return this._userAgent + ' ' + userAgent;
    }
    return this._userAgent;
  }

  _isUploadSpeedLimit(body: Readable, limit: number): Readable {
    if (limit == 0) {
      return body;
    }
    //Todo: finish speed limit readstream
    return body;
  }

  _ifRange(header: any): boolean {
    if (header && header['Range']) {
      return true;
    }
    return false;
  }

  _headerCast<T>(obj: any, clazz: T): { [key: string]: string } {
    let ret: { [key: string]: string } = {};
    let clz = clazz as any;
    let names: { [key: string]: string } = clz.names();
    Object.keys(names).forEach((key) => {
      let targetName = names[key];
      let value = obj[key];
      if (typeof value === 'undefined' || value == null) {
        ret[targetName] = '';
        return;
      }
      ret[targetName] = value.toString();
    });
    return ret;
  }

  _toHeader(header: any): { [key: string]: string } {
    let ret: { [key: string]: string } = {};
    if (!header) {
      return ret;
    }
    if (header.constructor && typeof header.constructor.names == 'function') {
      return this._headerCast(header, header.constructor);
    }
    for (let [key, value] of Object.entries(header)) {
      if (typeof value === 'undefined' || value == null) {
        ret[key] = '';
        continue;
      }
      ret[key] = value.toString();
    }
    return ret;
  }

  _readAsStream(response: $tea.Response): Readable {
    return response.body;
  }

  _getTracker(): any {
    throw new Error('the method is un-implemented!');
  }

  async _getAccessKeyID(): Promise<string> {
    return await this._creadentials.getAccessKeyId();
  }

  async _getAccessKeySecret(): Promise<string> {
    return await this._creadentials.getAccessKeySecret();
  }

  async _getSecurityToken(): Promise<string> {
    return await this._creadentials.getSecurityToken();
  }

  _getErrMessage(xml: string): { [key: string]: any } {
    let body: { [key: string]: any } = parseXML(xml);
    return body.Error || {};
  }

  _toXML(body: any): string {
    const builder = new Builder();
    return builder.buildObject(body);
  }

  _notNull(obj: {[key: string]: any}): boolean {
    return !!obj;
  }

  _inject(readable: Readable, ref: { [key: string]: string }): Readable {
    let shasum = createHash('md5');
    let crc64 = new Crc64();
    readable.on('data', function (chunk: Buffer) {
      shasum.update(chunk);
      crc64.update(chunk);
    });
    readable.on('end', function () {
      ref['md5'] = shasum.digest('base64');
      ref['crc'] = crc64.sum();
    });
    readable.pause();
    return readable;
  }
}

File: baseclient/go/service/baseclient.go
package service

import (
	"crypto/md5"
	"encoding/base64"
	"encoding/json"
	"encoding/xml"
	"errors"
	"fmt"
	"hash/crc64"
	"io"
	"net/http"
	"net/url"
	"reflect"
	"runtime"
	"strconv"
	"strings"
	"time"

	"github.com/alibabacloud-go/tea/tea"
	teautil "github.com/alibabacloud-go/tea/utils"
	"github.com/aliyun/alibabacloud-oss-sdk/baseclient/go/utils"
	"github.com/aliyun/credentials-go/credentials"
)

var crcTable = func() *crc64.Table {
	return crc64.MakeTable(crc64.ECMA)
}

var defaultUserAgent = fmt.Sprintf("AlibabaCloud (%s; %s) Golang/%s Core/%s", runtime.GOOS, runtime.GOARCH, strings.Trim(runtime.Version(), "go"), "0.01")

// Client is for calling oss api
type BaseClient struct {
	RegionId         string   `json:"RegionId" xml:"RegionId"`
	Protocol         string   `json:"Protocol" xml:"Protocol"`
	Endpoint         string   `json:"Endpoint" xml:"Endpoint"`
	UserAgent        string   `json:"UserAgent" xml:"UserAgent"`
	HostModel        string   `json:"HostModel" xml:"HostModel"`
	SignatureVersion string   `json:"SignatureVersion" xml:"SignatureVersion"`
	IsEnableMD5      bool     `json:"IsEnableMD5" xml:"IsEnableMD5"`
	IsEnableCrc      bool     `json:"IsEnableCrc" xml:"IsEnableCrc"`
	ReadTimeout      int      `json:"ReadTimeout" xml:"ReadTimeout"`
	ConnectTimeout   int      `json:"ConnectTimeout" xml:"ConnectTimeout"`
	LocalAddr        string   `json:"LocalAddr" xml:"LocalAddr"`
	HttpProxy        string   `json:"HttpProxy" xml:"HttpProxy"`
	HttpsProxy       string   `json:"HttpsProxy" xml:"HttpsProxy"`
	NoProxy          string   `json:"NoProxy" xml:"NoProxy"`
	Socks5Proxy      string   `json:"SOCKS5Proxy" xml:"SOCKS5Proxy"`
	Socks5NetWork    string   `json:"SOCKS5NetWork" xml:"SOCKS5NetWork"`
	MaxIdleConns     int      `json:"MaxIdleConns" xml:"MaxIdleConns"`
	AddtionalHeaders []string `json:"AddtionalHeaders" xml:"AddtionalHeaders"`
	credential       credentials.Credential
	Logger           *teautil.Logger
}

// ServiceError is for recording error which is caused when sending request
type ServiceError struct {
	Code      string `json:"Code" xml:"Code"`
	Message   string `json:"Message" xml:"Message"`
	RequestId string `json:"RequestId" xml:"RequestId"`
	HostId    string `json:"HostId" xml:"HostId"`
}

// New a client according to config
func (client *BaseClient) InitClient(config map[string]interface{}) error {
	client.Protocol = getStringValue(config["protocol"])
	client.RegionId = getStringValue(config["regionId"])
	client.Endpoint = getStringValue(config["endpoint"])
	client.UserAgent = getStringValue(config["userAgent"])
	client.HostModel = getStringValue(config["hostModel"])
	client.SignatureVersion = getStringValue(config["signatureVersion"])
	client.IsEnableMD5 = getBoolValue(config["isEnableMD5"])
	client.IsEnableCrc = getBoolValue(config["isEnableCrc"])
	client.ReadTimeout = getIntValue(config["readTimeout"])
	client.ConnectTimeout = getIntValue(config["connectTimeout"])
	client.LocalAddr = getStringValue(config["localAddr"])
	client.HttpProxy = getStringValue(config["httpProxy"])
	client.HttpsProxy = getStringValue(config["httpsProxy"])
	client.NoProxy = getStringValue(config["noProxy"])
	client.Socks5Proxy = getStringValue(config["socks5Proxy"])
	client.Socks5NetWork = getStringValue(config["socks5NetWork"])
	client.MaxIdleConns = getIntValue(config["maxIdleConns"])
	conf := &credentials.Configuration{
		AccessKeyID:     getStringValue(config["accessKeyId"]),
		AccessKeySecret: getStringValue(config["accessKeySecret"]),
		SecurityToken:   getStringValue(config["securityToken"]),
		Type:            getStringValue(config["type"]),
	}
	if conf.Type == "" {
		conf.Type = "access_key"
	}
	credential, err := credentials.NewCredential(conf)
	client.credential = credential
	if err != nil {
		return err
	}

	return nil
}

// func (client *BaseClient) SetLogger(level string, channel string, out io.Writer, template string) {
// 	client.Logger = teautil.NewLogger(level, channel, out, template)
// }

func (client *BaseClient) ReadAsStream(response *tea.Response) io.ReadCloser {
	return response.Body
}

func (client *BaseClient) GetUserAgent() string {
	if client.UserAgent == "" {
		return defaultUserAgent
	}
	return defaultUserAgent + " " + client.UserAgent
}

func (client *BaseClient) Empty(val string) bool {
	return val == ""
}

func (client *BaseClient) Equal(val, val2 string) bool {
	return val == val2
}

func (client *BaseClient) IfListEmpty(val []string) bool {
	return len(val) == 0
}

func (client *BaseClient) ReadAsString(resp *tea.Response) (string, error) {
	byt, err := resp.ReadBody()
	if err != nil {
		return "", err
	}
	return string(byt), err
}

func (client *BaseClient) GetSignatureV1(request *tea.Request, bucketName, accessKeySecret string) string {
	resource := ""
	if bucketName != "" {
		resource = "/" + bucketName
	}
	resource = resource + request.Pathname
	if !strings.Contains(resource, "?") {
		resource += "?"
	}
	for key, value := range request.Query {
		if isParamSign(key) {
			if value != "" {
				if strings.HasSuffix(resource, "?") {
					resource = resource + key + "=" + value
				} else {
					resource = resource + "&" + key + "=" + value
				}
			}
		}
	}
	return getSignedStrV1(request, resource, accessKeySecret)
}

func (client *BaseClient) GetSignatureV2(request *tea.Request, bucketName, accessKeySecret string, additionalHeaders []string) string {
	resource := ""
	pathName := request.Pathname
	if bucketName != "" {
		pathName = "/" + bucketName + request.Pathname
	}

	strs := strings.Split(pathName, "?")
	resource += uriEncode(strs[0], true)
	hs := newSorter(request.Query)
	if strings.Contains(pathName, "?") {
		hs.Keys = append(hs.Keys, strs[1])
		hs.Vals = append(hs.Vals, "")
	}

	// Sort the temp by the ascending order
	hs.Sort()
	if len(hs.Keys) > 0 {
		resource += "?"
	}
	for i := range hs.Keys {
		if !strings.HasSuffix(resource, "?") {
			resource += "&"
		}
		if hs.Vals[i] == "" {
			resource += uriEncode(hs.Keys[i], true)
		} else {
			resource += uriEncode(hs.Keys[i], true) + "=" + uriEncode(hs.Vals[i], true)
		}
	}
	return getSignedStrV2(request, resource, accessKeySecret, additionalHeaders)
}

// Parse string to uint
func (client *BaseClient) ParseUint(respCrc string, hasRange bool) uint64 {
	if hasRange {
		return 0
	}
	crc, err := strconv.ParseUint(respCrc, 10, 64)
	if err != nil {
		return 0
	}
	return crc
}

// func (client *BaseClient) IsUploadSpeedLimit(body io.Reader, limitSpeed int) io.Reader {
// 	if limitSpeed == 0 {
// 		return body
// 	}
// 	uploadLimiter := GetOssLimiter(limitSpeed)
// 	return &LimitSpeedReader{
// 		reader:     body,
// 		ossLimiter: uploadLimiter,
// 	}
// }

// If num is not 0, return num, or return defaultNum
func (client *BaseClient) DefaultNumber(num int, defaultNum int) int {
	if num != 0 {
		return num
	}
	return defaultNum
}

func (client *BaseClient) ToHeader(raw map[string]interface{}) map[string]string {
	byt, _ := json.Marshal(raw)
	header := make(map[string]string)
	err := json.Unmarshal(byt, &header)
	if err != nil {
		return nil
	}
	return header
}

// Parse filter to produce a map[string]string
func (client *BaseClient) ToQuery(filter map[string]interface{}) map[string]string {
	result := make(map[string]string)
	for key, value := range filter {
		filterValue := reflect.ValueOf(value)
		flatRepeatedList(filterValue, result, key)
	}
	return result
}

// Get Date in GMT
func (client *BaseClient) GetDate() string {
	return time.Now().UTC().Format(http.TimeFormat)
}

// Add prefix to key of meta
func (client *BaseClient) ToMeta(meta map[string]string, prefix string) map[string]string {
	result := make(map[string]string)
	for key, value := range meta {
		if !strings.HasPrefix(strings.ToLower((key)), prefix) {
			key = prefix + key
		}
		result[key] = value
	}
	return result
}

// Get value from obj according to key
func (client *BaseClient) GetSpecialValue(obj map[string]interface{}, key string) string {
	tmp := make(map[string]string)
	byt, _ := json.Marshal(obj)
	err := json.Unmarshal(byt, &tmp)
	if err != nil {
		return ""
	}

	return tmp[key]
}

// Remove prefix from key of meta
func (client *BaseClient) ParseMeta(meta map[string]string, prefix string) map[string]string {
	userMeta := make(map[string]string)
	for key, value := range meta {
		if strings.HasPrefix(strings.ToLower(key), prefix) {
			key = strings.Replace(key, prefix, "", 1)
		}
		userMeta[key] = value
	}
	return userMeta
}

func (client *BaseClient) GetErrMessage(bodyStr string) map[string]interface{} {
	resp := make(map[string]interface{})
	errMsg := &ServiceError{}
	err := xml.Unmarshal([]byte(bodyStr), errMsg)
	if err != nil {
		return resp
	}
	resp["Code"] = errMsg.Code
	resp["Message"] = errMsg.Message
	resp["RequestId"] = errMsg.RequestId
	resp["HostId"] = errMsg.HostId
	return resp
}

// Parse Body to map[string]interface{}
func (client *BaseClient) ParseXml(val string, result interface{}) map[string]interface{} {
	resp := make(map[string]interface{})

	start := getStartElement([]byte(val))
	out, err := XmlUnmarshal([]byte(val), result)
	if err != nil {
		return resp
	}
	resp[start] = out
	return resp
}

// Determine whether the request failed
func (client *BaseClient) IsFail(response *tea.Response) bool {
	return response.StatusCode < 200 || response.StatusCode >= 300
}

// If realStr is not "", return realStr, or return defaultStr
func (client *BaseClient) Default(realStr string, defaultStr string) string {
	if realStr != "" {
		return realStr
	}
	return defaultStr
}

// Return md5 according to body
func (client *BaseClient) GetContentMD5(a string) string {
	if !client.IsEnableMD5 {
		return ""
	}

	sum := md5.Sum([]byte(a))
	b64 := base64.StdEncoding.EncodeToString(sum[:])
	return b64
}

func (client *BaseClient) ToXML(obj map[string]interface{}) string {
	return mapToXML(obj)
}

func (client *BaseClient) NotNull(obj map[string]interface{}) bool {
	return len(obj) > 0
}

// Return content-type according to object name
func (client *BaseClient) GetContentType(name string) string {
	return utils.TypeByExtension(name)
}

func (client *BaseClient) GetAccessKeyID() (string, error) {
	if client.credential == nil {
		return "", errors.New("No credential valid!")
	}
	accessKeyId, err := client.credential.GetAccessKeyId()
	return accessKeyId, err
}

func (client *BaseClient) GetAccessKeySecret() (string, error) {
	if client.credential == nil {
		return "", errors.New("No credential valid!")
	}
	accessKeySecret, err := client.credential.GetAccessKeySecret()
	return accessKeySecret, err
}

func (client *BaseClient) ListToString(a []string, sep string) string {
	return strings.Join(a, sep)
}

func (client *BaseClient) GetSecurityToken() (string, error) {
	if client.credential == nil {
		return "", errors.New("No credential valid!")
	}
	token, err := client.credential.GetSecurityToken()
	return token, err
}

// Encryption
func (client *BaseClient) Encode(val string, encodeType string) string {
	strs := strings.Split(val, "/")
	if encodeType == "Base64" {
		encodeStr := base64.StdEncoding.EncodeToString([]byte(strs[len(strs)-1]))
		strs[len(strs)-1] = encodeStr
	} else if encodeType == "UrlEncode" {
		encodeStr, err := url.QueryUnescape(strs[len(strs)-1])
		if err != nil {
			return ""
		}
		strs[len(strs)-1] = encodeStr
	}
	return strings.Join(strs, "/")
}

// Decryption
func (client *BaseClient) Base64Decode(value string) string {
	strs := strings.Split(value, "/")
	result, err := base64.StdEncoding.DecodeString(strs[len(strs)-1])
	if err != nil {
		return ""
	}
	strs[len(strs)-1] = string(result)
	return strings.Join(strs, "/")
}

// Decryption
func (client *BaseClient) UrlDecode(value string) string {
	strs := strings.Split(value, "/")
	result, err := url.QueryUnescape(strs[len(strs)-1])
	if err != nil {
		return ""
	}
	strs[len(strs)-1] = result
	return strings.Join(strs, "/")
}

func (client *BaseClient) Inject(body io.Reader, ref map[string]string) io.Reader {
	body = ComplexReader(body, ref)
	return body
}
